File: src/movieclaw_api/schemas/jobs.py
"""持久化后台任务的公共 API 契约。"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from movieclaw_db.models import JobStatus


class JobProgressView(BaseModel):
    mode: str = Field(description="determinate / indeterminate / waiting / paused")
    phase: str = Field(description="当前领域阶段的稳定标识")
    message: str = Field(description="面向用户的即时进展")
    current: int | None = Field(default=None, description="当前完成量")
    total: int | None = Field(default=None, description="总量；未知时为空")
    percent: float | None = Field(default=None, description="真实可计算时才返回百分比")
    phase_index: int | None = Field(default=None, description="当前阶段序号")
    phase_count: int | None = Field(default=None, description="总阶段数")
    details: dict[str, Any] = Field(default_factory=dict, description="领域进度明细")


class JobResourceView(BaseModel):
    resource_type: str
    resource_id: str
    relation: str


class JobView(BaseModel):
    id: str
    job_type: str
    subject: str | None
    definition_version: int
    handler_revision: str
    prompt_revision: str | None
    provider_ref: str | None
    status: JobStatus
    input_data: dict[str, Any] = Field(description="已确认的任务输入（永不包含密钥）")
    progress: JobProgressView
    result: dict[str, Any] | None
    error: dict[str, Any] | None
    usage: dict[str, Any]
    resources: list[JobResourceView] = Field(default_factory=list)
    origin: str
    actor_kind: str | None
    actor_name: str | None
    parent_job_id: str | None
    root_job_id: str | None
    retry_of_job_id: str | None
    attempt: int
    max_attempts: int
    revision: int
    cancel_requested_at: datetime | None
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None
    finished_at: datetime | None


class JobListView(BaseModel):
    items: list[JobView]


class JobWaitView(BaseModel):
    changed: bool
    job: JobView


class JobCancelView(BaseModel):
    cancelled: bool
    job: JobView


class JobRetryView(BaseModel):
    created: bool
    job: JobView


class JobEventView(BaseModel):
    id: int
    job_id: str
    revision: int
    event_type: str
    payload: dict[str, Any]
    created_at: datetime


class JobEventListView(BaseModel):
    items: list[JobEventView]


class JobWorkerHealthView(BaseModel):
    running: bool
    owner: str | None
    active_workers: int
    counts: dict[str, int]
    oldest_queued_seconds: int | None

File: apps/web/lib/api/jobs.ts
import { request } from "@/lib/http";

interface ApiEnvelope<T> {
  success: boolean;
  code: string;
  message: string;
  data: T;
}

export type JobStatus =
  | "queued"
  | "running"
  | "retry_wait"
  | "cancelling"
  | "waiting"
  | "blocked"
  | "succeeded"
  | "failed"
  | "cancelled";

export interface JobProgress {
  mode: "determinate" | "indeterminate" | "waiting" | "paused";
  phase: string;
  message: string;
  current: number | null;
  total: number | null;
  percent: number | null;
  phase_index: number | null;
  phase_count: number | null;
  details: Record<string, unknown>;
}

export interface JobResource {
  resource_type: string;
  resource_id: string;
  relation: string;
}

export interface JobView {
  id: string;
  job_type: string;
  subject: string | null;
  definition_version: number;
  handler_revision: string;
  prompt_revision: string | null;
  provider_ref: string | null;
  status: JobStatus;
  input_data: Record<string, unknown>;
  progress: JobProgress;
  result: Record<string, unknown> | null;
  error: {
    code?: string;
    message?: string;
    details?: Record<string, unknown>;
    actions?: Array<{ type: string; label: string; target?: string }>;
  } | null;
  usage: Record<string, unknown>;
  resources: JobResource[];
  origin: string;
  actor_kind: string | null;
  actor_name: string | null;
  parent_job_id: string | null;
  root_job_id: string | null;
  retry_of_job_id: string | null;
  attempt: number;
  max_attempts: number;
  revision: number;
  cancel_requested_at: string | null;
  /** 取消发起方；`system:` 前缀代表系统自动收口，重跑没有意义，界面不给「重新执行」。 */
  cancel_requested_by: string | null;
  created_at: string;
  updated_at: string;
  started_at: string | null;
  finished_at: string | null;
}

/** 系统自动收口的取消（如整树入库完成后清理被取代的分批任务）：重跑没有意义，不提供「重新执行」。 */
export function isSystemCancelled(job: JobView): boolean {
  return job.status === "cancelled" && (job.cancel_requested_by?.startsWith("system:") ?? false);
}

export const ACTIVE_JOB_STATUSES = new Set<JobStatus>([
  "queued",
  "running",
  "retry_wait",
  "cancelling",
  "waiting",
  "blocked",
]);

export async function listJobs(options: {
  activeOnly?: boolean;
  limit?: number;
} = {}): Promise<JobView[]> {
  const query = new URLSearchParams({ limit: String(options.limit ?? 100) });
  if (options.activeOnly) query.set("active_only", "true");
  const response = await request<ApiEnvelope<{ items: JobView[] }>>(`/jobs?${query}`);
  return response.data.items;
}

export async function cancelJob(jobId: string): Promise<JobView> {
  const response = await request<ApiEnvelope<{ cancelled: boolean; job: JobView }>>(
    `/jobs/${jobId}/cancel`,
    { method: "POST" },
  );
  return response.data.job;
}

export async function retryJob(jobId: string): Promise<JobView> {
  const response = await request<ApiEnvelope<{ created: boolean; job: JobView }>>(
    `/jobs/${jobId}/retry`,
    { method: "POST" },
  );
  return response.data.job;
}
